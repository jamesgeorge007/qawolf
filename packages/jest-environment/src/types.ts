import { Browser } from "@qawolf/browser";
import { Runner } from "@qawolf/runner";
import { BrowserStep, ScrollValue, Workflow } from "@qawolf/types";
import { Page } from "puppeteer";

declare global {
  // declare the globals we expose in RunnerEnvironment
  const runner: Runner;

  function click(step: BrowserStep): Promise<void>;
  function input(step: BrowserStep, value?: string | null): Promise<void>;
  function scroll(step: BrowserStep, value: ScrollValue): Promise<void>;

  const steps: BrowserStep[];
  const values: (string | undefined)[];
  const workflow: Workflow;

  const browser: Browser;
  function currentPage(): Promise<Page>;
  function getPage(
    index?: number,
    waitForRequests?: boolean,
    timeoutMs?: number
  ): Promise<Page>;
}
