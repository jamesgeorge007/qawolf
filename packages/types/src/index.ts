export type Callback<S = void, T = void> = (data?: S) => T;

export type Action = "click" | "input" | "scroll";

export type BrowserStep = {
  action: Action;
  index: number;
  pageId?: number;
  target: ElementDescriptor;
  value?: string | ScrollValue | null;
};

export type ElementDescriptor = {
  ariaLabel?: string | null;
  classList?: string[] | null;
  dataValue?: string | null;
  href?: string | null;
  iconContent?: string[] | null;
  id?: string | null;
  inputType?: string | null;
  labels?: string[] | null;
  name?: string | null;
  parentText?: string[] | null;
  placeholder?: string | null;
  tagName?: string | null;
  textContent?: string | null;
  title?: string | null;
  xpath?: string | null;
};

export interface Event {
  action: Action;
  isTrusted: boolean;
  pageId?: number;
  target: ElementDescriptor;
  time: number;
}

export interface InputEvent extends Event {
  action: "input";
  value?: string | null;
}

export type ScrollValue = {
  x: number;
  y: number;
};

export interface ScrollEvent extends Event {
  action: "scroll";
  value: ScrollValue;
}

export type Locator = {
  action: Action;
  dataAttribute: string | null;
  target: ElementDescriptor;
  timeoutMs: number;
  value?: string | null;
};

export type Size = "desktop" | "tablet" | "mobile";

export type Workflow = {
  name: string;
  size: Size;
  steps: BrowserStep[];
  url: string;
};
