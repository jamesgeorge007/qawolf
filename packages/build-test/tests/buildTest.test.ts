import { BrowserStep } from "@qawolf/types";
// import directly since fixtures are not exported
import { loginWorkflow } from "../../build-workflow/fixtures/loginWorkflow";
import {
  buildTest,
  formatIt,
  formatMethod,
  formatStep
} from "../src/buildTest";

describe("buildTest", () => {
  it("builds a test from a workflow", async () => {
    const testString = buildTest(loginWorkflow);
    expect(testString).toMatchSnapshot();
  });
});

describe("formatIt", () => {
  it("formats labels", () => {
    const step: BrowserStep = {
      action: "input",
      index: 0,
      target: {
        labels: ["name", "username"],
        name: "other",
        tagName: "input"
      }
    };

    expect(formatIt(step)).toBe('can input "name username" input');
  });
});

describe("formatMethod", () => {
  it("throws error if invalid step action", () => {
    expect(() => {
      formatMethod("drag" as any, 0);
    }).toThrowError();
  });
});

describe("formatStep", () => {
  it("formats click on link step", () => {
    const step: BrowserStep = {
      action: "click",
      index: 0,
      target: {
        tagName: "a",
        textContent: "contact"
      }
    };

    const formattedStep = formatStep(step);
    expect(formattedStep).toMatchSnapshot();
  });

  it("formats click on submit input step", () => {
    const step: BrowserStep = {
      action: "click",
      index: 1,
      target: {
        inputType: "submit",
        tagName: "input",
        textContent: "sign in"
      }
    };

    const formattedStep = formatStep(step);
    expect(formattedStep).toMatchSnapshot();
  });

  it("formats scroll action", () => {
    const step: BrowserStep = {
      action: "scroll",
      index: 0,
      target: { xpath: "/html" },
      value: { x: 0, y: 10 }
    };

    const formattedStep = formatStep(step);
    expect(formattedStep).toMatchSnapshot();
  });

  it("formats type into text input", () => {
    const step: BrowserStep = {
      action: "input",
      index: 0,
      target: {
        id: "input1",
        inputType: "text",
        labels: ["username"],
        name: "user",
        placeholder: "Jane Doe",
        tagName: "input"
      },
      value: "spirit"
    };

    const formattedStep = formatStep(step);
    expect(formattedStep).toMatchSnapshot();
  });

  it("formats type into password input", () => {
    const step: BrowserStep = {
      action: "input",
      index: 10,
      target: {
        id: "input2",
        inputType: "password",
        placeholder: "secret",
        tagName: "input"
      },
      value: "supersecret"
    };

    const formattedStep = formatStep(step);
    expect(formattedStep).toMatchSnapshot();
  });

  it("shortens target name if needed", () => {
    const step: BrowserStep = {
      action: "click",
      index: 0,
      target: {
        inputType: "submit",
        tagName: "input",
        textContent: `sign in${"x".repeat(200)}`
      }
    };

    const formattedStep = formatStep(step);
    expect(formattedStep).toMatchSnapshot();
  });
});
