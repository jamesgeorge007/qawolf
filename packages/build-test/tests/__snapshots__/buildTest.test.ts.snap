// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildTest builds a test from a workflow 1`] = `
"describe('login', () => {
  it('can scroll', async () => {
    await scroll(steps[0], values[0]);
  });

  it('can click \\"form authentication\\" link', async () => {
    await click(steps[1]);
  });

  it('can input into \\"username\\" input', async () => {
    await input(steps[2], values[2]);
  });

  it('can click \\"login\\" i', async () => {
    await click(steps[3]);
  });

  it('can input into \\"username\\" input', async () => {
    await input(steps[4], values[4]);
  });

  it('can input into \\"password\\" input', async () => {
    await input(steps[5], values[5]);
  });

  it('can click \\"login\\" i', async () => {
    await click(steps[6]);
  });
});"
`;

exports[`formatStep formats click on link step 1`] = `
Object {
  "it": "can click \\"contact\\" link",
  "method": "click(steps[0])",
}
`;

exports[`formatStep formats click on submit input step 1`] = `
Object {
  "it": "can click \\"sign in\\" input",
  "method": "click(steps[1])",
}
`;

exports[`formatStep formats scroll action 1`] = `
Object {
  "it": "can scroll",
  "method": "scroll(steps[0], values[0])",
}
`;

exports[`formatStep formats type into password input 1`] = `
Object {
  "it": "can input into \\"secret\\" input",
  "method": "input(steps[10], values[10])",
}
`;

exports[`formatStep formats type into text input 1`] = `
Object {
  "it": "can input into \\"username\\" input",
  "method": "input(steps[0], values[0])",
}
`;

exports[`formatStep shortens target name if needed 1`] = `
Object {
  "it": "can click \\"sign inxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx...\\" input",
  "method": "click(steps[0])",
}
`;
